### PR TITLE
Lecture 4 solutions

### PR DESCRIPTION
**Pull request by racakenon**
### Solutions for Lecture 4

cc @chshersh

### CODE REVIEW

**Comment by racakenon** [comment]
lecture 2 eval function refactoring

```haskell
eval env exp = case exp of
  Lit i -> Right i
  Var x -> case lookup x env of
    Nothing -> Left (VariableNotFound x)
    Just v -> Right v
  Add expl expr -> eval env expl >>= \l -> eval env expr >>= \r -> Right (l + r)
```

File: src/Lecture4.hs [hunk 1]
```diff
@@ -134,8 +137,24 @@ errors. We will simply return an optional result here.
...
+  n <- readMaybe c
+  guard $ n >= 0
+  pure n
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow l = do
+  [prod, trade, cost] <- Just $ split l
+  guard $ prod /= ""
+  readMaybe trade >>= \t -> parseCost cost >>= \c -> pure $ Row prod t c
```
Comment (chshersh) [inline src/Lecture4.hs:157]: Awesome implementation using the `MonadFail` instance for `Maybe` and `do`-notation? 👏🏻

**Review by chshersh — APPROVED** [review]
:tada: Fantastic work! Congratulations on finishing all **four** lectures! 🥳 
You're truly a Functional Adept now 🧠 
I hope you had fun :man_dancing: :microphone:

If you would like to receive a pretty certificate about the course completion, please, send your full name and the link to this PR to `haskell.beginners2022@gmail.com` 📧 

If you have time, I would appreciate your feedback on the course:

* [Haskell Beginners 2022 Course Feedback](https://docs.google.com/forms/d/e/1FAIpQLSfCkST64dnDCYkz5mezkVyLfWjeskSxhV0Y_cYxUaJw0U9ZhQ/viewform)

If you like my work, you can [sponsor me on GitHub](https://github.com/sponsors/chshersh) 💖

And if you want to follow my work in the future (books, courses, blogging, etc.), you can subscribe to [my mailing list](kodimensional.dev/#subscribe) 📫 

You can also ask any questions here or anywhere else and we can discuss any topics you would like to dig into as well  :slightly_smiling_face: Don't be a stranger!

**Comment by chshersh** [comment]
@racagogi I don't have much to add, your solution already looks great 🙂

You can improve it if you want to work on making the implementation properly lazy and streaming but besides that, the code looks rather idiomatic 👍🏻 

On a separate note, indeed, you can rewrite `eval` from a previous lecture much easier, and you can even use the standard function `liftA2` to write the code even shorter:

```diff
-Add expl expr -> eval env expl >>= \l -> eval env expr >>= \r -> Right (l + r)
+Add expl expr -> liftA2 (+) (eval env expl) (eval env expr)
```

**Pull request merged**